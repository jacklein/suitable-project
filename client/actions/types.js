// badge actions
export const FETCH_BADGES = 'fetch_badges';
export const UPDATE_PROGRESS = 'update_progress';