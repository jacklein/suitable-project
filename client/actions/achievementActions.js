import {
  FETCH_ACHIEVEMENTS
} from './types';
import axios from 'axios';

export const fetchAchievements = () => async dispatch => {
  const res = await axios.get('http://192.168.1.137:3001/v1/achievements');
  dispatch({ type: FETCH_ACHIEVEMENTS, payload: res.data });
}