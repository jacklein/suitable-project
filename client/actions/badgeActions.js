import {
  FETCH_BADGES, 
  UPDATE_PROGRESS
} from './types';
import axios from 'axios';

const URI = 'http://192.168.1.137:3001'

/**
 * res.data structure:
 * [
 * { 'id': 123, 'details': {...}, 'relationships': {...} },
 * { 'id': 456, 'details': {...}, 'relationships': {...} } 
 * ]
 */
export const fetchBadges = () => async dispatch => {
  try{
    const res = await axios.get(`${URI}/v1/achievements`);
    dispatch({ type: FETCH_BADGES, payload: res.data });
  } catch(e) {
    console.log(e);
  }
}

/**
 * res.data structure:
 * {
 * 'id': (id of completed activity),
 * 'updates': [{'achievementID': 456, 'progress': 0.5}, {...}, {...}]
 * }
 */
export const updateProgress = (id, callback) => async dispatch => {
  try {
    const res = await axios.get(`${URI}/v1/activities/${id}/complete`);
    dispatch({ type: UPDATE_PROGRESS, payload: res.data.updates });
    callback();
  } catch(e) {
    console.log(e);
  }
}