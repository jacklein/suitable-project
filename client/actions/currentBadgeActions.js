import {
  SET_INDEX
} from './types';

/**
 * Sets the index of the badge currently being looked at
 */
export const setIndex = index => async dispatch => {
  dispatch({ type: SET_INDEX, payload: index });
}