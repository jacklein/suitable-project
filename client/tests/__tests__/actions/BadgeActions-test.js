import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import moxios from 'moxios';
import { fetchBadges, updateProgress } from '../../../actions/BadgeActions';
import { achievements, updates } from '../../mockdata';
import { FETCH_BADGES, UPDATE_PROGRESS } from '../../../actions/types';

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

describe('BadgeActions', () => {

  beforeEach(function () {
    moxios.install();
  });

  afterEach(function () {
    moxios.uninstall();
  });

  it('creates a FETCH_BADGES action after successfully fetching achievements', () => {
    moxios.wait(() => {
      const request = moxios.requests.mostRecent();
      request.respondWith({
        status: 200,
        response: achievements,
      });
    });

    const expectedActions = [
      { type: FETCH_BADGES, payload: achievements },
    ];

    const store = mockStore({ badges: [] });

    return store.dispatch(fetchBadges()).then(() => {
      // return of async actions
      expect(store.getActions()).toEqual(expectedActions);

      expect(store.getActions()).toMatchSnapshot();
    });
  });

  it('creates an UPDATE_PROGRESS action after successfully fetching updated progresses', () => {
    moxios.wait(() => {
      const request = moxios.requests.mostRecent();
      request.respondWith({
        status: 200,
        response: updates,
      });
    });

    const expectedActions = [
      { type: UPDATE_PROGRESS, payload: updates.updates },
    ];

    const store = mockStore({ badges: [] });

    return store.dispatch(updateProgress({ id: '123', callback: () => {} })).then(() => {
      // return of async actions
      expect(store.getActions()).toEqual(expectedActions);

      expect(store.getActions()).toMatchSnapshot();
    });
  });
});
