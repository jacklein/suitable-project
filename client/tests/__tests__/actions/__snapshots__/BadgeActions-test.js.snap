// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BadgeActions creates a FETCH_BADGES action after successfully fetching achievements 1`] = `
Array [
  Object {
    "payload": Array [
      Object {
        "details": Object {
          "description": "Sit nisi eveniet non expedita eos. Porro et odit vel quis voluptatem error. Et est autem nostrum quis voluptatem voluptas voluptatibus sint. Occaecati nesciunt provident dolorem tenetur rerum officiis.",
          "iconUrl": "https://www.filepicker.io/api/file/h6spFMWTQaKXzk6HOxqL",
          "progress": 0.29699748744715526,
          "title": "Exceptional Achievement",
        },
        "id": "1924883835",
        "relationships": Object {
          "recommendations": Array [
            Object {
              "details": Object {
                "description": "Fuga id aspernatur sed pariatur itaque placeat. Velit unde aut explicabo. Pariatur fuga eveniet ipsa. Aut amet repudiandae laboriosam doloremque sit. Ut quam molestiae ab occaecati illo sed unde. Harum minus est libero eaque cum.",
                "title": "Bacon 21151",
                "type": "TASK",
              },
              "id": "1056306996",
            },
            Object {
              "details": Object {
                "description": "Sit et maiores reiciendis amet voluptate voluptas totam dolores et. Ut ipsum nostrum soluta. In ut suscipit nihil distinctio molestias ut quia sint dignissimos. Cum hic perferendis deleniti ipsum totam dolor ad id. Qui minima non harum ut iste maiores doloremque libero.",
                "title": "Bedfordshire 31245",
                "type": "EVENT",
              },
              "id": "2492513005",
            },
            Object {
              "details": Object {
                "description": "Tempore dolore excepturi quia officia et illo culpa voluptatum. Dolor maiores blanditiis animi. Doloremque earum tempore ex expedita velit ab. Odio nostrum necessitatibus eaque nesciunt voluptatem nam est. Non laboriosam dicta assumenda.",
                "title": "Manager 349",
                "type": "EVENT",
              },
              "id": "1036388801",
            },
            Object {
              "details": Object {
                "description": "Ut qui neque ex facere et. Aperiam illum sapiente ipsa tenetur. Sequi perspiciatis fugit illum animi eos quia cupiditate fugiat. Recusandae ea beatae ipsam.",
                "title": "Village 81370",
                "type": "EVENT",
              },
              "id": "2945439988",
            },
            Object {
              "details": Object {
                "description": "Illo asperiores id pariatur. Qui voluptatem id quaerat dicta. Facere nobis facilis et voluptatem sed veritatis.",
                "title": "coherent 27723",
                "type": "EVENT",
              },
              "id": "2217784005",
            },
          ],
        },
      },
    ],
    "type": "fetch_badges",
  },
]
`;

exports[`BadgeActions creates an UPDATE_PROGRESS action after successfully fetching updated progresses 1`] = `
Array [
  Object {
    "payload": Array [
      Object {
        "achievementId": "1924883835",
        "progress": 0.6656102509337432,
      },
    ],
    "type": "update_progress",
  },
]
`;
